20250429-RS-scheduler-broadcast-job-type.md

```diff
@@ -0,0 +1,320 @@
+
+#### Scheduler
+
+The Scheduler service will be updated to include a new Job handler code path to build a Broadcast Job manager.
+This manager will take informer events from go-etcd-cron, sending PUT and DELETE flagged job events to connected clients during CRUD events, as well as when appropriate clients connect.
+The Scheduler will trigger all appropriate connected clients concurrently during trigger events.
+It is not expected that clients would return any response code to a trigger execution other than SUCCESS, as these jobs are generally intended (at least for the initial use cases) to provide object manifests- not triggering some execution.
+A client failing to return the response of a trigger should be considered as a failed connection.
+In any case, to account for future use cases, the trigger response codes should considered a logical AND whereby failed calls will be retired on each failed client call according to the Jobs Failure Policy.
+
+Broadcast jobs can either expire according to its schedule, or be deleted by a client.
+
+Broadcast Jobs will continue to adhere to the staging queue, like existing Jobs today.
+
+#### Keep Alive
+
+The gRPC keep alive options for both the Scheduler server and daprd clients need to be updated so that they are more aggressive.
+This ensures greater durability of the gRPC job streams for Scheduler broadcast job manager.
+https://pkg.go.dev/google.golang.org/grpc/keepalive
+
+#### protos
+
+The following protos will be added to the Scheduler service to implement the Broadcast job type.
+Note that both the Actor Subscription and Durable Actor ID feature use cases are included as example of API shape.
+
+```proto
+// JobTargetType is the type of the job target.
+enum JobTargetType {
+  // JOB_TARGET_TYPE_JOB is the job target type.
+  JOB_TARGET_TYPE_JOB = 0;
+
+  // JOB_TARGET_TYPE_ACTOR_REMINDER is the actor reminder target type.
+  JOB_TARGET_TYPE_ACTOR_REMINDER = 1;
+
+  // JOB_TARGET_TYPE_ACTOR_SUBSCRIPTION is the broadcast actor subscription
+  // object target type.
+  JOB_TARGET_TYPE_ACTOR_SUBSCRIPTION = 2;
+
+  // JOB_TARGET_TYPE_DURABLE_ACTOR_ID is the broadcast durable actor ID target type.
+  JOB_TARGET_TYPE_DURBALE_ACTOR_ID = 3;
+}
+
+// JobTargetMetadata holds the typed metadata associated with the job for
+// different origins.
+message JobTargetMetadata {
+  oneof type {
+    TargetJob job = 1;
+    TargetActorReminder actor_reminder = 2;
+    TargetBroadcast broadcast = 3;
+  }
+}
+```
+
+```proto
+syntax = "proto3";
+
+package dapr.proto.scheduler.v1;
+
+import "google/protobuf/any.proto";
+
+option go_package = "github.com/dapr/dapr/pkg/proto/scheduler/v1;scheduler";
+
+// BroadcastJobEventType is the type of the event why this job is being
+// broadcast.
+enum BroadcastJobEventType {
+  // BROADCAST_JOB_EVENT_PUT is the event type for a job being created or
+  // updated.
+  BROADCAST_JOB_EVENT_PUT = 0;
```

Comment:
I'd propose that we favor immutability from the get-go and treat the job event registrations as immutable, requiring a delete and recreate instead of allowing updates.

The use of some sort of one-shot `AddOrReplace` method would be useful here so the operation can be done from the caller in one operation (avoid race or inconsistency conditions).